js-packages/@prestojs/rest/src/viewModelCachingMiddleware.ts

```diff
@@ -0,0 +1,197 @@
+ * single object or an array of objects)
+ *
+ * ```js
+ * const middleware = viewModelCachingMiddleware({
+ *     "records.users": User,
+ *     "records.bookings": Booking,
+ * });
+ * ```
+ *
+ * The mapping can optionally be a function that returns a mapping or returns a promise that
+ * resolves to a mapping. This is useful for dealing with dynamic imports or returning a class
+ * that isn't yet defined.
+ *
+ * ```js
+ * const middleware = paginationMiddleware(() => {
+ *   const Booking = (await import('./Booking')).default;
+ *   return {
+ *     "records.users": User,
+ *     "records.bookings": Booking,
+ *   }
+ * });
+ * ```
+ *
+ * Each record instance created is also automatically added to the cache.
+ *
+ * NOTE: If using with [paginationMiddleware](doc:paginationMiddleware) then this must come
+ * before `paginationMiddleware`.
+ *
+ * @param viewModelMapping The mapping to use for caching as described above
+ * @param options
+ *
+ * @extract-docs
+ * @menu-group Middleware
+ */
+export default function viewModelCachingMiddleware<ReturnT = any>(
+    viewModelMapping: ViewModelMappingDef,
+    options: ViewModelCachingOptions<ReturnT> = {}
+): MiddlewareObject<ReturnT> {
+    const { getDeleteId = defaultGetDeleteId } = options;
+    const resolveViewModelMapping = async (): Promise<ViewModelMapping> =>
+        !isViewModelClass(viewModelMapping) && typeof viewModelMapping === 'function'
+            ? viewModelMapping()
+            : viewModelMapping;
+    const cacheAndTransform = async (data: any): Promise<any> => {
+        const _viewModelMapping = await resolveViewModelMapping();
+
+        if (isViewModelClass(_viewModelMapping)) {
+            return cacheDataForModel(_viewModelMapping, data);
+        }
+        const transformed = cloneDeep(data);
+        Object.entries(_viewModelMapping).forEach(([key, viewModel]) => {
+            const value = get(data, key);
+            if (value !== undefined) {
+                set(transformed, key, cacheDataForModel(viewModel, value));
+            }
+        });
+        return transformed;
+    };
+
+    return {
+        contributeToClass(endpoint: Endpoint): void {
+            // When using the default implementation we can check things are setup correctly on initialisation and
+            // throw an error. For custom implementations we have to wait until the method is called to do the check.
+            if (getDeleteId === defaultGetDeleteId && endpoint.requestInit.method === 'DELETE') {
+                if (!endpoint.urlPattern.requiredArgNames.includes('id')) {
+                    throw new Error(
+                        `When using 'viewModelCachingMiddleware' on a DELETE endpoint it is expected the UrlPattern includes an 'id' parameter. Known parameters are: ${endpoint.urlPattern.validArgNames.join(
+                            ', '
+                        )}. You can pass 'getDeleteId' to override this behavior.`
+                    );
```

Comment:
it does but it's about providing better dev experience over purity

i kinda don't think it's worth making it more complex - if you are overriding getDeleteId you are probably doing it for a specific endpoint and wouldn't be worth writing the code to do that check

i guess if your project used a different convention altogether it might be wroth it

Reply:
ok this was easy enough just to add a `validateEndpoint` function so i've moved the check in there